File: web/src/core/errors/AppError.ts
/**
 * Standardized application error codes across the domain and API boundary.
 */
export type AppErrorCode =
  | 'INVALID_INPUT'
  | 'BOOKING_CODE_NOT_FOUND'
  | 'STALE_SELECTIONS'
  | 'UPSTREAM_BETWAY_ERROR'
  | 'INTERNAL_SERVER_ERROR';

/**
 * Standard error response envelope payload.
 */
export interface AppErrorPayload {
  code: AppErrorCode;
  message: string;
  details: unknown | null;
}

/**
 * Standard error response envelope.
 */
export interface AppErrorResponse {
  success: false;
  error: AppErrorPayload;
}

const DEFAULT_STATUS_CODES: Record<AppErrorCode, number> = {
  INVALID_INPUT: 400,
  BOOKING_CODE_NOT_FOUND: 404,
  STALE_SELECTIONS: 422,
  UPSTREAM_BETWAY_ERROR: 502,
  INTERNAL_SERVER_ERROR: 500,
};

const DEFAULT_MESSAGES: Record<AppErrorCode, string> = {
  INVALID_INPUT: 'Invalid input provided.',
  BOOKING_CODE_NOT_FOUND: 'The provided Betway booking code could not be found or has expired.',
  STALE_SELECTIONS: 'One or more selections are no longer active or available.',
  UPSTREAM_BETWAY_ERROR: 'Failed to communicate with Betway upstream service.',
  INTERNAL_SERVER_ERROR: 'An unexpected internal server error occurred.',
};

/**
 * Canonical Application Error class for domain, gateway, and route handler exceptions.
 */
export class AppError extends Error {
  public readonly code: AppErrorCode;
  public readonly statusCode: number;
  public readonly details: unknown | null;

  constructor(
    code: AppErrorCode,
    message?: string,
    details?: unknown,
    statusCode?: number
  ) {
    const finalMessage = message ?? DEFAULT_MESSAGES[code] ?? 'Application error';
    super(finalMessage);
    this.name = 'AppError';
    this.code = code;
    this.statusCode = statusCode ?? DEFAULT_STATUS_CODES[code] ?? 500;
    this.details = details !== undefined ? details : null;

    // Maintain prototype chain
    Object.setPrototypeOf(this, AppError.prototype);
  }

  /**
   * Converts the error instance into a JSON serializable error payload.
   */
  public toPayload(): AppErrorPayload {
    return {
      code: this.code,
      message: this.message,
      details: this.details,
    };
  }

  /**
   * Converts the error instance into the standardized API error response envelope.
   */
  public toResponse(): AppErrorResponse {
    return {
      success: false,
      error: this.toPayload(),
    };
  }

  // --- Static Factory Helpers ---

  public static invalidInput(message?: string, details?: unknown): AppError {
    return new AppError('INVALID_INPUT', message, details);
  }

  public static notFound(message?: string, details?: unknown): AppError {
    return new AppError('BOOKING_CODE_NOT_FOUND', message, details);
  }

  public static staleSelections(message?: string, details?: unknown): AppError {
    return new AppError('STALE_SELECTIONS', message, details);
  }

  public static upstreamError(message?: string, details?: unknown): AppError {
    return new AppError('UPSTREAM_BETWAY_ERROR', message, details);
  }

  public static internal(message?: string, details?: unknown): AppError {
    return new AppError('INTERNAL_SERVER_ERROR', message, details);
  }

  /**
   * Translates any caught error into a canonical AppError.
   */
  public static from(err: unknown): AppError {
    if (err instanceof AppError) {
      return err;
    }

    if (err instanceof Error) {
      return new AppError('INTERNAL_SERVER_ERROR', err.message, { stack: err.stack });
    }

    return new AppError('INTERNAL_SERVER_ERROR', String(err));
  }
}

File: mobile/lib/domain/models/app_error.dart
import 'package:equatable/equatable.dart';

/// Canonical application error codes across the domain and gateway boundary.
abstract class AppErrorCodes {
  static const String invalidInput = 'INVALID_INPUT';
  static const String bookingCodeNotFound = 'BOOKING_CODE_NOT_FOUND';
  static const String staleSelections = 'STALE_SELECTIONS';
  static const String upstreamBetwayError = 'UPSTREAM_BETWAY_ERROR';
  static const String networkError = 'NETWORK_ERROR';
  static const String internalServerError = 'INTERNAL_SERVER_ERROR';
}

/// Canonical Domain Exception / Error class.
/// Decoupled from transport libraries (Dio/Retrofit) and presentation layers.
class AppError extends Equatable implements Exception {
  /// Machine-readable error code (e.g. 'BOOKING_CODE_NOT_FOUND')
  final String code;

  /// Human-readable error message
  final String message;

  /// HTTP status code if originated from backend / transport
  final int? statusCode;

  /// Optional contextual details
  final dynamic details;

  const AppError({
    required this.code,
    required this.message,
    this.statusCode,
    this.details,
  });

  /// Factory for invalid input errors (HTTP 400).
  factory AppError.invalidInput({
    String message = 'Invalid input provided.',
    dynamic details,
  }) {
    return AppError(
      code: AppErrorCodes.invalidInput,
      message: message,
      statusCode: 400,
      details: details,
    );
  }

  /// Factory for not found errors (HTTP 404).
  factory AppError.notFound({
    String message =
        'The provided Betway booking code could not be found or has expired.',
    dynamic details,
  }) {
    return AppError(
      code: AppErrorCodes.bookingCodeNotFound,
      message: message,
      statusCode: 404,
      details: details,
    );
  }

  /// Factory for stale selection errors (HTTP 422).
  factory AppError.staleSelections({
    String message =
        'One or more selections are no longer active or available.',
    dynamic details,
  }) {
    return AppError(
      code: AppErrorCodes.staleSelections,
      message: message,
      statusCode: 422,
      details: details,
    );
  }

  /// Factory for upstream Betway errors (HTTP 502).
  factory AppError.upstreamError({
    String message = 'Failed to communicate with Betway upstream service.',
    dynamic details,
    int? statusCode,
  }) {
    return AppError(
      code: AppErrorCodes.upstreamBetwayError,
      message: message,
      statusCode: statusCode ?? 502,
      details: details,
    );
  }

  /// Factory for network / connectivity errors.
  factory AppError.networkError({
    String message =
        'Network connection failed. Please check your internet connection.',
    dynamic details,
  }) {
    return AppError(
      code: AppErrorCodes.networkError,
      message: message,
      details: details,
    );
  }

  /// Factory for internal / unhandled errors (HTTP 500).
  factory AppError.internal({
    String message = 'An unexpected internal error occurred.',
    dynamic details,
    int? statusCode,
  }) {
    return AppError(
      code: AppErrorCodes.internalServerError,
      message: message,
      statusCode: statusCode ?? 500,
      details: details,
    );
  }

  @override
  String toString() =>
      'AppError(code: $code, message: $message, statusCode: $statusCode)';

  @override
  List<Object?> get props => [code, message, statusCode, details];
}
